### DISCUSSION TITLE
Progress messages

### DISCUSSION THREAD

**Discussion by WhoIsSethDaniel** [bscan/PerlNavigator#53] [Ideas]
Any chance for progress messages? Even something as simple as 'I am initializing', 'I am done initializing', 'I am formatting', 'I am done formatting' would be great. 

The reason for this is that it isn't always clear, upon start, when the server is ready to start doing things. Also, for formatting (and similar tasks) I don't know if the server is doing something until suddenly my code changes. This can take awhile with `perltidy` and `perlimports`.

**Comment by bscan** [comment]
Hi Seth. I'm curious how long the Navigator is taking on startup. Normally, I'd expect it to be pretty fast. Below is a gif of me loading vscode on my somewhat slow Surface. It takes a bit for vscode to startup, load the file, and apply syntax highlighting, but relatively soon afterwards the Navigator starts and produces compilation errors. There's not much to do for initialization, so I believe most of the delay is waiting for the editor to start the server. Once it's running, it should be ready to respond to requests almost immediately. It runs a startup task to list all modules available for autocompletion, but this runs asynchronously, so the Navigator should be able to compile files in the meantime (while having slightly degraded autocompletion). Also, I only restart my editor a few times per week at max. I generally always leave it open, so I don't think about startup time much. Startup gif below: 

![vscode](https://user-images.githubusercontent.com/10503608/205516096-909618d1-df08-45f3-bd75-060e4dd042c6.gif)

And for Format document, I just timed it out. For an arbitrary 1000 line file (I used List::MoreUtils::PP) It takes about 1 second to perltidy and about 2 seconds to run perl imports. VSCode provides a visual indicator of an outstanding request, which is helpful in this case.
For editors that don't provide this, I'm not really sure where I could put a progress message. One possible speed-up could be splitting perltidy and perlimports into different calls. In the Python LSP, they have "Format Document" distinct from "Sort Imports". That would reduce the latency of the request. Sorting gif below

![vscodeTiming](https://user-images.githubusercontent.com/10503608/205516445-421d3af8-4b9e-40ac-a1f2-75ff9f75fa76.gif)

**Reply by WhoIsSethDaniel** [reply]
I'm not a vscode user so my experience is with neovim. Regardless, I don't think it matters that much which editor one uses.

Re: initialization: In some cases I can start neovim with a perl file, navigate to something, ask for 'go-to-definition' and it won't happen because the server is (presumably) still initializing. It's not a lot of time, but it is enough time for this to happen. The only feedback I get that the server is now ready is if there is a `perlcritic` error to display. Once the error is displayed I know the LSP server has completed initialization and I can continue. 

Re: formatting: I don't actually use the built-in `perltidy` due to #11. I turn off `perltidy` support, but I have `perlimports` turned on. `perlimports` can be quite slow. At work there are numerous perl files where it will take 5+ seconds to do its job. It's often not obvious that it has completed its job since there is no feedback. `perltidy` can also take more than just 1 or 2 seconds, although I  admit it's pretty rare. Although if you're on a VM with a slow disk even small tasks seem to take a second or two longer. 

For what it's worth there are other LSP servers that send progress messages regardless of how long it is expected to take for an operation. An example would be Go's LSP server `gopls`. Nearly everything takes less than a second, particularly formatting, yet it sends progress messages. This is useful because I now know a) it is running and b) it has finished. Without those messages I wouldn't necessarily know that anything had happened. (`gopls` also takes a very long time to initialize on even modestly sized projects so it also helps there). 

Consider, also, that not everyone is developing on a speedy, local host. I often am compelled to develop on a slow(er) VM in a remote location. So while these operations are fast on a local SSD they may be much slower on a VM with slow disk access.

**Reply by bscan** [reply]
Hi @WhoIsSethDaniel! Thanks for advocating for this idea, and I'm convinced by your argument. It certainly seems far better to get progress messages than have the Navigator appear frozen or non-functional. I just committed a first version as version 0.4.3 on NPM: https://github.com/bscan/PerlNavigator/commit/adbb50acf5e40138ddeba8ae46498f3092ae445b 

By default, the progress messages are turned off, but you can enable them with the setting `enableProgress`. Before I can enable it by default, I'll also need to check `capabilities` since I believe not all clients have support for progress markers. Also, worth testing to make sure they work well across a few different editors.

With the Navigator, each file needs to first be compiled to get go-to definition results. The progress message should now show the first time a file is compiled. It's not sending any messages on server initialization, but I hope the file compilation message is sent early enough in the process. Progress messages should also be sent while document formatting. Let me know how it works for you, and I'll test in some other editors as well. Thanks!

**Comment by WhoIsSethDaniel** [comment]
Hi @bscan. I have yet to change anything but the version of the server, but ATM the server is crashing. Often it seems to crash not long after start. I'll continue to look at it and see if this is a problem on my end.

**Reply by WhoIsSethDaniel** [reply]
With 0.4.3 this happens if I run the code from the command-line:
```
node:internal/modules/cjs/loader:988
  throw err;
  ^

Error: Cannot find module 'nanoid/non-secure'
Require stack:
- /home/sdaniel_maxmind_com/.local/share/nvim/mason/packages/perlnavigator/node_modules/perlnavigator-server/out/progress.js
- /home/sdaniel_maxmind_com/.local/share/nvim/mason/packages/perlnavigator/node_modules/perlnavigator-server/out/formatting.js
- /home/sdaniel_maxmind_com/.local/share/nvim/mason/packages/perlnavigator/node_modules/perlnavigator-server/out/server.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:985:15)
    at Function.Module._load (node:internal/modules/cjs/loader:833:27)
    at Module.require (node:internal/modules/cjs/loader:1057:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/home/sdaniel_maxmind_com/.local/share/nvim/mason/packages/perlnavigator/node_modules/perlnavigator-server/out/progress.js:5:22)
    at Module._compile (node:internal/modules/cjs/loader:1155:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
    at Module.load (node:internal/modules/cjs/loader:1033:32)
    at Function.Module._load (node:internal/modules/cjs/loader:868:12)
    at Module.require (node:internal/modules/cjs/loader:1057:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/sdaniel_maxmind_com/.local/share/nvim/mason/packages/perlnavigator/node_modules/perlnavigator-server/out/progress.js',
    '/home/sdaniel_maxmind_com/.local/share/nvim/mason/packages/perlnavigator/node_modules/perlnavigator-server/out/formatting.js',
    '/home/sdaniel_maxmind_com/.local/share/nvim/mason/packages/perlnavigator/node_modules/perlnavigator-server/out/server.js'
  ]
}
```
This does not happen when I reinstall 0.4.1.

**Reply by bscan** [reply]
Thank you for testing, and my apologies for the crash. Nanoid is a dependency I added for generating random tokens for tracking progress events. Appears that I forgot to add it to package.json. That's odd that it still worked for me though in my testing. I can fix that this evening.

**Reply by WhoIsSethDaniel** [reply]
Thanks for doing the work! I look forward to testing this when you get the new package out.

**Reply by bscan** [reply]
The newest version 0.4.4 is up on NPM and included nanoid as a dependency: https://www.npmjs.com/package/perlnavigator-server?activeTab=dependencies

**Reply by WhoIsSethDaniel** [reply]
I am seeing progress messages now. I see both the initialization message(s) and formatting messages. For formatting I am only seeing what appears to be a completed formatting progress message and nothing prior to that. I need to see what's being sent to the client before I can say more.

**Reply by WhoIsSethDaniel** [reply]
It does seem that the server is sending both the begin and end formatting progress indicator one right after the other.  I mentioned above that I don't use the `perltidy` integration, but I do use the `perlimports` integration. When I format the file I see a long delay before the formatting begin/end come in. I wonder if `perlimports` is running during that delay. This makes me wonder if progress messages are being sent for `perlimports`? Or if they are being sent for `perltidy` despite it being turned off? Is this possible?

**Reply by WhoIsSethDaniel** [reply]
Even if I turn on `perltidy` support and add a `sleep` into `perltidy` the begin/end messages only seem to appear after the formatting has occurred.

**Reply by bscan** [reply]
Could be related to the formatting function being synchronous instead of async, and how different clients deal with them. The code is fairly short. Start progress, run perlimports and perltidy, and then endsProgress: https://github.com/bscan/PerlNavigator/blob/main/server/src/formatting.ts 

How about the file initialization progress? Do they display well, and provide meaningful feedback? They're also using the same progress functions which worked in my testing in vscode and the Kate editor (although I did more testing in initialization messages than formatting)

**Reply by WhoIsSethDaniel** [reply]
Initialization works perfectly. I see the "Initializing &lt;file&gt;" message. There is a spinner (via the UI) that spins until the "finish" message is received at which point a check-mark appears in-place of the spinner.

**Reply by WhoIsSethDaniel** [reply]
I did look at the new code prior to posting earlier. My best guess is that there is a buffering issue. But that's as detailed as I can go at this point. :-(  If I knew more about TS or JS I might be able to fix it myself (assuming the problem is on the server side).

**Reply by bscan** [reply]
Nice. I'm glad the file initialization part is working. That's a good start. The formatting is using the same code, and is the same usage other than the sync vs async function difference. Currently formatting is a synchronous operation, so you won't be able to go-to definition or do other operations while your code is formatting. Originally, this seemed like a reasonable restriction, but is likely getting in the way. The code compiling and critic have always been async since they need to regularly run in the background re-checking files.

Looking through the vscode-languager-server source here: https://github.com/microsoft/vscode-languageserver-node/blob/main/client/src/common/client.ts, sendProgress is being sent asynchronously (I'm using the vscode libraries to build the server, even for non-vscode editors).  However, the messages aren't sent until later because formatting is blocking the event loop. I'll see if I can change the formatting process to run async as well (or force the progress messages to resolve synchronously). Thanks for your patience, sounds like we're almost there.

**Reply by WhoIsSethDaniel** [reply]
I commented out the two lines of progress code in `formatting.js/maybeReturnEdits` and placed them in `server.js/onDocumentFormatting` and the formatting progress messages worked as expected. This was a hack not a fix, but I think points to what the problem is/may be.

**Reply by WhoIsSethDaniel** [reply]
I just saw your reply. It seems you have a handle on the situation. Thanks again for doing this work.

**Reply by bscan** [reply]
Just pushed another version (0.4.5). Hopefully third time's the charm 😆 . I changed the necessary functions to be asynchronous, and the progress messages seem to be working. Let me know. Thanks!

**Reply by WhoIsSethDaniel** [reply]
I still have the same issue. :frowning_face:

**Reply by WhoIsSethDaniel** [reply]
@bscan Much like when I modified the generated code in https://github.com/bscan/PerlNavigator/discussions/53#discussioncomment-4346596 I did something similar here. And, again, it works. But this modification is a quick hack.  What did I do?
I removed the `startProgress` and `endProgress` lines from `out/formatting.js/maybeReturnEdits` and placed them in `out/server.js/connection.onDocumentFormatting`. That function now looks like this:
```javascript
 connection.onDocumentFormatting(async (params) => {
     let document = documents.get(params.textDocument.uri);
     let settings = documentSettings.get(params.textDocument.uri);
     const progressToken = await (0, progress_1.startProgress)(connection, 'Formatting doc', settings);
     const workspaceFolders = await getWorkspaceFoldersSafe();
     if (!document || !settings)
         return;
     console.log(params);
     const editOut = await (0, formatting_1.formatDoc)(params, document, settings, workspaceFolders, connection);
     (0, progress_1.endProgress)(connection, progressToken);
     return editOut;
 });
```
My only guess as to why this works is the extra `async` operation when `getWorkspaceFoldersSafe` is called.

**Reply by bscan** [reply]
Ah, this is a tricky one. My guess is that the extra async operations allow more time for the progress messages to be sent before the entire event loop is blocked by the formatting operation. I should probably just change the entire formatting operation to be async. Still not clear why this works on some editors and not others, although perhaps different clients process messages at different rates. I'll see if I can get this done this weekend.

**Comment by bscan** [comment]
Hi @WhoIsSethDaniel , I just got back to this now that #11 has been fixed. I suspect progress messages are even more important if waiting for both perltidy and perlimports to finish.  I changed them both to run asynchronously via https://github.com/bscan/PerlNavigator/commit/99e7c2d375bf940f7c1e473a5237662ce50c699d . Thanks for the suggestions! Hopefully this change gets it working

**Comment by WhoIsSethDaniel** [comment]
it works well. Thanks!
